module: expose exports conditions to loaders

jkrems opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

jkrems commented:
/cc @nodejs/modules-active-members

jkrems commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -19,7 +19,28 @@ const { ERR_INPUT_TYPE_NOT_ALLOWED,
 
-function defaultResolve(specifier, { parentURL } = {}, defaultResolve) {
+const DEFAULT_CONDITIONS_KEY = 'node\0import';
+const DEFAULT_CONDITION_SET = new Set(['node', 'import']);
```
I think it may be better to have this cache as a `std::set<std::string>` within the C++ code and only generate the key here. `default` is on purpose not included. I think it makes sense if loaders cannot influence it.

addaleax commented on src/module_wrap.cc:
```diff
@@ -932,6 +933,7 @@ Maybe<URL> ResolveExportsTarget(Environment* env,
                                 const std::string& subpath,
                                 const std::string& pkg_subpath,
                                 const URL& base,
+                                Local<Set>& conditions,
```
```suggestion
                                Local<Set> conditions,
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1058,7 +1060,8 @@ Maybe<bool> IsConditionalExportsMainSugar(Environment* env,
                               const PackageConfig& pcfg,
-                              const URL& base) {
+                              const URL& base,
+                              Local<Set>& conditions) {
```
```suggestion
                              Local<Set> conditions) {
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1115,7 +1118,8 @@ Maybe<URL> PackageExportsResolve(Environment* env,
                                  const PackageConfig& pcfg,
-                                 const URL& base) {
+                                 const URL& base,
+                                 Local<Set>& conditions) {
```
```suggestion
                                 Local<Set> conditions) {
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1176,7 +1180,8 @@ Maybe<URL> PackageExportsResolve(Environment* env,
                        const std::string& pkg_subpath,
-                       const URL& base) {
+                       const URL& base,
+                       Local<Set>& conditions) {
```
```suggestion
                       Local<Set> conditions) {
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1206,7 +1211,8 @@ Maybe<URL> ResolveSelf(Environment* env,
                           const std::string& specifier,
-                          const URL& base) {
+                          const URL& base,
+                          Local<Set>& conditions) {
```
```suggestion
                          Local<Set> conditions) {
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1294,7 +1300,8 @@ Maybe<URL> PackageResolve(Environment* env,
                    const std::string& specifier,
-                   const URL& base) {
+                   const URL& base,
+                   Local<v8::Set>& conditions) {
```
```suggestion
                   Local<Set> conditions) {
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -991,12 +993,12 @@ Maybe<URL> ResolveExportsTarget(Environment* env,
                          key->ToString(context).ToLocalChecked());
       std::string key_str(*key_utf8, key_utf8.length());
-      if (key_str == "node" || key_str == "import") {
+      if (conditions->Has(context, key).FromJust()) {
```
You need to bubble exceptions up to JS rather than using `.FromJust()` for operations on JS sets (because these *can* fail in the presence of termination exceptions as they may use V8-internal JS code – and of course it’s also good practice to properly handle exceptions)

jkrems replied:
Updated to do a `To(&bool)` and return an empty result. Afaict that should be enough to preserve the proper exception bubbling?

addaleax commented on src/module_wrap.cc:
```diff
@@ -1330,10 +1337,15 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
   }
 
+  CHECK(args[2]->IsSet());
+  Local<v8::Set> conditions = args[2].As<v8::Set>();
```
```suggestion
  Local<Set> conditions = args[2].As<Set>();
```

addaleax commented on src/module_wrap.cc:
```diff
@@ -1330,10 +1337,15 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
 
+  CHECK(args[2]->IsSet());
+  Local<v8::Set> conditions = args[2].As<v8::Set>();
+  Local<Context> context = args.GetIsolate()->GetCurrentContext();
```
Is this unused?
```suggestion
```

jkrems replied:
It is! Thanks for catching this.

guybedford commented:
Nice work, this approach seems great!

GeoffreyBooth commented on test/fixtures/es-module-loaders/loader-with-custom-condition.mjs:
```diff
@@ -0,0 +1,6 @@
+export async function resolve(specifier, options, next) {
```
Can we please use the same variable names for function arguments for all the hooks in the tests? I found it difficult to update all the tests with new signatures when I was working on my PR because the tests’ variable names were previously all over the place (`parentURL`, `parentModuleURL`, `parentModuleUrl`, etc.). I think these should just copy whatever we put in the docs; and if those change, then all the tests change.

```js
export async function resolve(specifier, context, defaultResolve) {
```

jkrems replied:
Not worth bike-shedding here so I just used those parameter names.

jkrems commented:
Cleaned up and marking as ready for review. I updated the implementation to use a `std::set` instead of `V8::Set`. In the process the conditions cache got moved into C++, JS only creates the key (`\0`-separated condition names).

guybedford commented on src/module_wrap.cc:
```diff
@@ -1364,10 +1377,34 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+      conditions_cache_
+        .insert(std::make_pair(conditions_key, parsed_conditions))
+        .first;
+  }
```
I think this code would actually be simpler and shorter if it were the direct JS array being iterated, like the Object keys iterations in this file?

Or is there a reason you didn't want to pass in a userland JS array here?

guybedford replied:
If the issue is error handling, it should be fine to abort here though still I think?

A test case for that might be something like - 

```js
defaultResolve(specifier, { ...context, conditions: Object.defineProperty(['a'], '0', { get () { throw new Error('blah') } }) })
```

or something like that.

jkrems replied:
The question is: Do we want to cache the set or do we want to reconstruct it each time? My idea here is that most of the time (custom loaders absent), the code path is just "get the existing set from this string" without ever needing an array.

We could still forward the array as an additional argument instead of parsing the key. But that seems like an opportunity for things to get out of sync..? We could create a safe array from the key (`.split('\n')`) on the JS side but that seems wasteful.

I don't think we have to care about getters because the conditions array will likely be serialized and sent through message channels. So in a final design I don't think getters would survive until this layer (or at least they shouldn't be supported).

jkrems replied:
I can update this to accept both the key and the array but I'm not sure if that makes it a whole lot easier since we trade the string split for a bunch of iteration+type assertion+cast-to-native-string boilerplate.

guybedford replied:
I would still prefer to see this change if possible, but not a blocker.

jkrems replied:
I'd be happy to discuss in a follow-up PR but I'd rather keep the C++ side simple and handle edge cases on the JS side where there's clearer stack traces etc..

guybedford commented on test/fixtures/es-module-loaders/loader-with-custom-condition.mjs:
```diff
@@ -0,0 +1,6 @@
+  const customContext = Object.assign({}, context, {
+    conditions: ['custom-condition'].concat(context.conditions),
+  });
+  return defaultResolve(specifier, customContext);
```
`return defaultResolve(specifier, { ...context, conditions: ['custom-condition', ...context.conditions] })` might be simpler.

jkrems replied:
Updated. This also showed that `conditions` wasn't provided to the loader, updated that as well. The loader now gets a frozen default conditions array.

jkrems commented:
Ping @nodejs/modules-active-members, this is still lacking some reviews/approvals before we can land it. Adding it to the agenda in case it's still open by our next meeting. :)

GeoffreyBooth commented on test/fixtures/es-module-loaders/loader-with-custom-condition.mjs:
```diff
@@ -0,0 +1,11 @@
+export async function resolve(specifier, context, defaultResolve) {
+  ok(Array.isArray(context.conditions), 'loader receives conditions array');
+  ok(context.conditions.includes('node'), 'provided conditions include "node"');
+  ok(!context.conditions.includes('default'), 'default is implicit');
```
So is `context.conditions` always an object with keys `node` and `import` (per `const DEFAULT_CONDITIONS = ObjectFreeze(['node', 'import'])` above)? That seems . . . weird. Why not `import` and `require` if you’re going to do that, or `import` and `require` and `node` and `default` if we’re going to list all the conditions that Node itself supports?

Or just only list the conditions actually defined in the `package.json`? If conditional exports syntax isn’t used, I guess you would just do the one `default` condition (as not using the conditional syntax is the same as using it with the `default` condition, I think). And if conditional exports syntax _is_ used, include only the conditions that the user has defined.

jkrems replied:
> Why not import and require if you’re going to do that, or import and require and node and default if we’re going to list all the conditions that Node itself supports?

This code only runs (right now) for imports. So it shouldn't include `require` because then it would match `require` conditions for import. It lists all conditions that apply to this resolution request and can potentially be changed by the loader hook. I made the choice of not allowing the hook to mess with `default`.

If at some future point `require` resolution should pass through loader hooks, then those resolutions would have `conditions` set to `["node", "require"]` by default.

I'm not sure I follow the rest of your comment. The conditions are the ones that are active for this resolution. Nothing in any `package.json` will affect this. This list is potentially used to match the conditions in `package.json`. But there's no way to know which conditions will actually appear in any `package.json` file (or if any `package.json` file will be loaded in the first place) when the `resolve` hook is called.

GeoffreyBooth replied:
I'm confused. When you say that we're exposing the exports conditions, I assumed that what you meant was that for a particular `"exports"` left-hand-side, the entire right-hand-side object would be provided to the loader. Not just the particular condition that matched. Wouldn't we want all the information, so that the loader could choose on its own to pick a different condition? E.g. if someone made a `"typescript"` condition say, we'd want the loader to know that that exists and read it and choose to load that instead of the matched `"import"` or `"node"` or `"default"` condition.

bmeck replied:
This PR is to allow loaders to pass around a `"typescript"` condition; it doesn't affect resolution, but does allow custom loaders to alter the conditions that are being resolved thereby altering resolution by adding/removing/etc. the set of conditions.

I don't understand how the right-hand-side of `"exports"` comes into play here as the PR is only about the left hand side and it is up to loaders to resolve the right hand side.

GeoffreyBooth replied:
When I say left-hand-side, I'm referring to the path, e.g. the `"./feature"` in:

```js
  "exports": {
    "./feature": {
      "import": "./feature-default.js",
      "browser": "./feature-browser.js"
    }
```

bmeck replied:
Ah, this PR is about the "import"/"browser" LHS. We were talking about different things!

jkrems replied:
> E.g. if someone made a "typescript" condition say, we'd want the loader to know that that exists and read it and choose to load that instead of the matched "import" or "node" or "default" condition.

I think you're thinking about it from the opposite direction. The `resolve` comes before any `package.json` file has been loaded. There is no `exports` yet. There's only "what conditions are active for this resolution", not yet "what mappings are possible". So hook doesn't have to know the latter. If it wants to allow a `browser` or `typescript` mapping, it just needs to it to the conditions when it delegates to the parent hook.

Example realistic-ish hooks:

```js
export function resolve(specifier, context, next) {
  return next(specifier, {
    ...context,
    {
      conditions: [
        process.env.NODE_ENV === 'production' ? 'production' : 'development',
        ...context.conditions,
      ],
    }
  });
}
```

1. Custom resolve hook is called, it's from an import so we pass `["node", "import"]`.
2. The hook wants to add either `production` or `development` to the recognized conditions, so it calls the parent with those.
3. The built-in resolve logic searches for a `package.json` and will match using all the conditions passed by the hook.

I don't see how the hook could know about contents of the `package.json` and what kind of use case you are thinking about when you're talking about those keys.

A hook that uses the conditions itself, not just changes which are active:

```js
const CACHED_IMPORT_MAPS = loadImportMaps();

export function resolve(specifier, context, next) {
  const conditionsKey = context.conditions.slice(0).sort().join('\0');
  if (CACHED_IMPORT_MAPS.has(conditionsKey) {
    return resolveFromImportMap(
      specifier,
      context,
      CACHED_IMPORT_MAPS.get(conditionsKey)
    );
  }
  return next(specifier, context, next);
}
```

In neither of those cases it seems valuable to pass conditions extracted from a `package.json` to the hook.

GeoffreyBooth replied:
I'm not really understanding the purpose of this PR then. It would seem that the exposed conditions would always be `["node", "import"]`, since we currently don't have composition of loaders. In other words,

> The built-in resolve logic searches for a `package.json` and will match using all the conditions passed by the hook.

But then the Node default `resolve` would always ignore a `production` or `development` condition, so what purpose would have been served by adding such a custom condition? It would seem to be useful only if there was some other loader next in the chain that then would do something with this custom-added condition. Right? But there can't be one, at least not yet.

bmeck replied:
Node as of this PR would process all of the Set rather than just "import" and "node", thats what https://github.com/nodejs/node/pull/31303/files#diff-4a5950be44f56dfca2e7fa3c4e1fc0e4R1058 is changing

guybedford commented on test/fixtures/es-module-loaders/loader-with-custom-condition.mjs:
```diff
@@ -0,0 +1,11 @@
+
+export async function resolve(specifier, context, defaultResolve) {
+  ok(Array.isArray(context.conditions), 'loader receives conditions array');
+  ok(context.conditions.includes('node'), 'provided conditions include "node"');
```
What do you think about calling this `envConditions` for "environment conditions" rather than "conditions"? Just thinking in terms of terminology if we want to try and narrow this down? Or `nodeEnv` or even just `env`?

bmeck replied:
I'd avoid trying to state they are environmental, as I don't think that fits "require"

guybedford replied:
`require` and `import` are very specific exceptions to the rule. And can still sort of be considered "static" down to the fact that they are loader-environment-specific.

Overall conditions should be considered environment static though...

bmeck replied:
I disagree that they should be static and environmental

jkrems replied:
I think environment is already very overloaded, especially if dev/prod ever becomes one of the conditions, making (node) env a subset of conditions. Conditions also matches "conditional exports" and I think it's too late to rename that one.

guybedford replied:
What I'm arguing for here is two types of conditions:

1. "System conditions" - `"default"`, `"node"`, `"require"`, `"import"` (and later `"self"`) that cannot be changed or altered, since they apply based on their rules which are formally defined and to change those rules would be to break the contract.
2. "User environment conditions" - `process.env.NODE_ENV.split(',')` - that apply additionally to the system conditions, but not in any way that can interfere with them.

If we integrate a split like this, then I'd be happy to expose the `"user environment conditions"` to loaders through the context in this way, but I am against allowing loaders to break the contract of the system conditions in (1).

bmeck replied:
> since they apply based on their rules which are formally defined and to change those rules would be to break the contract.

Can you clarify this? I'm not sure what contract is being talked about.

guybedford replied:
Node.js should always match the `"node"` condition, since that is the definition of that condition. Packages can rely on this by using patterns like:

```js
{
  "exports": {
    "node": "./asdf.js",
    "default": "./DEFINITELY_NOT_NODEJS.js"
  }
}
```

and the contract in the above is that `"default"` will never ever be a Node.js main resolution.

I wouldn't want loaders to be able to just randomly change these set resolutions - but rather only to add new ones that don't have existing definitions for Node.js.

bmeck replied:
If a loader is meaning to emulate a different environment than Node I would expect it to be able to change the defaults to not include the "node" condition. I'm still unclear on the contract here, but it might be because I do not see loaders as being constrained to only emulate Node.

jkrems replied:
I think there's two different questions for me:

1. Should node allow loaders to mess with certain "reserved" conditions? We already restricted these by protecting `require`/`import` and not exposing `default`. We could expand this to the `node` conditions` (fail if the hook calls `defaultResolve` without it).
2. Should resolve hooks make assumptions about the conditions that are active? Should they hard-code that resolution includes `node` or `import` etc.?

Since I want to be able to use `resolve` hooks outside of node, I feel pretty strongly that (2) should *not* be the case. I would want to be able to reuse custom resolution hooks for non-node environments like bundlers. And the only way to allow that is if `node` is one of the arguments and not implicitly assumed by the hook.

guybedford replied:
Right, and that's why I think this PR is an over-generalization of a more simple `process.env.NODE_ENV` conditional approach.

We are trying to solve for perfect abstraction and the usability of users adding conditions to their apps. Those are two separate use cases.

bmeck commented on doc/api/esm.md:
```diff
@@ -1024,6 +1024,7 @@ this one, or `undefined` if this is the main entry point for the application.
  * @param {string} specifier
  * @param {object} context
  * @param {string} context.parentURL
+ * @param {string[]} context.conditions
```
can we add a note that passing `"require"` may be a breaking change in the future. If we end up causing it to route through the CJS resolver.

jkrems replied:
Should we disallow passing `require` right now to allow for that future possibility? Allowing import-style resolution but with `require` condition seems bad™.

bmeck replied:
I'm ok throwing for now, but that seems a bit perf heavy if you keep missing parse cache. I do think the resolve hook works fine for `require` design wise, but thats a separate discussion than this PR.

jkrems replied:
Added an error that is thrown when a loader tries to add `require` as a condition.

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -1024,6 +1024,7 @@ this one, or `undefined` if this is the main entry point for the application.
  * @param {string} specifier
  * @param {object} context
  * @param {string} context.parentURL
+ * @param {string[]} context.conditions
```
Sorry if I’m being dense, but I still don’t understand what this PR aims to achieve; _and I wrote much of the code for hooks,_ so I feel like I really should be getting this. At the very least, add some text to the documentation explaining what `context.conditions` is and what it can be used for.

jkrems replied:
The chain is:

1. Resolution starts within a specific context. The context includes who the parent is and certain meta information/tags about the resolution (the conditions). This includes "this is a resolution within node.js" (`node`) and "the resolution was triggered by an `import` statement or expression" (`import`).
2. The module loader calls the custom `resolve` hook with that information. It also passes the native resolution logic that does support conditional resolution as `defaultResolve` to the hook.
3. The custom `resolve` hook may inspect or modify the conditions. In the test that comes with this PR, it adds an additional condition for example (`custom-condition`). A more realistic hook may add `production` if it sees that `process.env.NODE_ENV` is set to `production`.
4. The custom `resolve` hook can call the `defaultResolve` with those modified conditions. In the test, that's exactly what it does.
5. `defaultResolve` then calls into the native C++ logic, passing the set of conditions that apply to the current resolution as provided by the custom hook.
6. When the C++ logic encounters a conditional export, it uses the set of active conditions to filter which conditions are matched. In the case of the included test, normally the C++ logic wouldn't match things guarded with `custom-condition`. But because the hook added that condition, it does now.

jkrems replied:
Will add a brief use to the example to give some context.

GeoffreyBooth commented:
I've been chatting with @jkrems about this. So I was confused by the PR title "expose exports conditions to loaders". The exports conditions aren't from `package.json` or any other user-defined source, but rather are _Node's_ supported conditions, which are currently always `"node"` and `"import"` where ES loaders are concerned. So this PR passes those in to custom `resolve` hooks via `context.conditions`. This isn't very useful, of course, as `context.conditions` is always those two values.

But that's not all! If the user then modifies that set of conditions and passes a modified conditions object into `defaultResolve`, this _changes the way the default ES loader behaves,_ to now resolve the specifier using these new conditions and priority order. So whereas previously a `"production"` condition in `package.json` would've been ignored, now it would be matched, without the need for a `getSource` hook. So what this PR is really doing is providing a way to configure the ES loader on the fly to support custom conditions, either globally or on a per-specifier basis. @jkrems, please correct me if I'm still misunderstanding any of the above.

I'm not sure about the returning an unknown value to `defaultResolve` changing how the default resolver works; it's quite different from how the other hooks behave. For example, if `getFormat` returns a format that's unknown to Node, Node doesn't do anything with it. Creating a `getFormat` hook that ends with `return defaultGetFormat(url, {format: 'foo'})` would throw an error (I think).

Anyway I'm still thinking about this. I agree with the overall goal of trying to support custom loading of custom conditions, e.g. a loader to enable different source code to be loaded based on a `NODE_ENV` value.

GeoffreyBooth commented:
I'm still thinking about this. So as I understand it, the goal of this PR is as follows:

- Allow custom conditions, e.g. `production` and `development`, to be used as part of resolution.
- Do so without excessive boilerplate.

The second part is key. A hook _could_ be written today, using `resolve` as currently shipped, that achieves the above goal. It would be something like (psuedocode):

```js
export async function resolve(specifier, context, defaultResolve) {
  if (specifier.startsWith('.') || specifier.startsWith('file://') {
    // Relative or absolute specifiers get default handling
    return defaultResolve(specifier, context, defaultResolve);
  } else {
    // Handle specifiers that start with a package name,
    // e.g. 'pkg' or 'pkg/foo' or 'pkg/file.js'
    const packageMetadata = getNearestParentPackageJsonFor(specifier);
    const exportsLeftHandSide = `.${specifier.substr(specifier.indexOf('/'))}`
    const exportsRightHandSide = packageMetadata.exports[exportsLeftHandSide];
    if (process.env.NODE_ENV === 'development' &&
        exportsRightHandSide.development) {
      // Get Node's resolved URL for 'pkg'
      const base = defaultResolve(packageMetadata.name);
      // Merge that with the exported path for the "development" condition
      return {
        url: new URL(exportsRightHandSide.development, base).href
      };
    }
  }
}
```

But longer, because you'd need to implement `getNearestParentPackageJsonFor` (which is similar to [this](https://github.com/nodejs/node/issues/30514#issuecomment-555201335)). So I see the problem.

But here's the thing: I think most `resolve` hooks that intend to do anything with bare specifiers or starting-with-bare specifiers (`'pkg'` or `'pkg/foo'` or `'pkg/file.js'`, for example) will probably need the `package.json` metadata for that bare specifier, e.g. `'pkg'`. So for such specifiers, why don't we look up the controlling `package.json` and add its contents to the context?

```js
export async function resolve(specifier, context, defaultResolve) {
  // Say specifier is 'pkg/feature'; because it doesn't start with '.' or 'file://',
  // context.packageMetadata is defined, and contains some or all of the contents
  // of the package.json file that Node *would* look up for 'pkg/feature'
  const { packageMetadata = null } = context;
  if (packageMetadata && packageMetadata.exports) {
    // packageMetadata is the parsed contents of ./node_modules/pkg/package.json
    const exportsLeftHandSide = `.${specifier.substr(specifier.indexOf('/'))}`;
    // exportsLeftHandSide is "./feature"
    const exportsRightHandSide = packageMetadata.exports[exportsLeftHandSide];
    if (exportsRightHandSide.development && process.env.NODE_ENV === 'development') {
      // Get Node's resolved URL for 'pkg'
      const base = defaultResolve(packageMetadata.name);
      // Merge that with the exported path for the "development" condition
      return {
        url: new URL(exportsRightHandSide.development, base).href
      };
    } else {
      return defaultResolve(specifier, context, defaultResolve);
    }
  } else {
    return defaultResolve(specifier, context, defaultResolve);
  }
}
```

In my opinion, this more closely follows the pattern of the other hooks: for a certain case, return this different value; else return default resolution. This also has the benefit that we can look at _any_ of the `package.json` metadata and potentially resolve anything based on it: want to resolve to what's in the `"module"` field? Go ahead!

The downside is that Node might potentially look up a `package.json` unnecessarily, if the custom hook doesn't actually use the supplied metadata. If we really care about this, we could make `context.packageMetadata` a getter that does `fs.readFileSync` only on read; or provide a function for getting the metadata. I'm also assuming a simplified `packageMetadata.exports`, that doesn't cover all the variations of the allowed syntax (array fallbacks, shorthand sugar, etc.). That's a problem that any tool that works with `"exports"` will face, which is its own problem that either will be addressed in userland via a package like `resolve` or that we can address via a utility function.

This API could even allow a user to follow the pattern that this PR takes with `context.conditions`: a hook could take `context.packageMetadata` and _modify it,_ passing the modified version into `defaultResolve`. This is essentially `context.conditions` on steroids, allowing the hook to rewrite anything about what got returned from `package.json`; though without the part of this PR where passed-in unknown conditions suddenly get supported. I'm still very skeptical of that aspect of this PR.

```js
export async function resolve(specifier, context, defaultResolve) {
  const { packageMetadata = null } = context;
  if (packageMetadata && packageMetadata.exports) {
    const exportsLeftHandSide = `.${specifier.substr(specifier.indexOf('/'))}`
    const exportsRightHandSide = packageMetadata.exports[exportsLeftHandSide];
    if (exportsRightHandSide.development && process.env.NODE_ENV === 'development') {
      // Overwrite all conditions with our new value
      packageMetadata.exports[exportsLeftHandSide] = exportsRightHandSide.development;
      // Now get Node to resolve things as if this had been the package.json
      context.packageMetadata = packageMetadata;
      return defaultResolve(specifier, context, defaultResolve);
    } else {
      return defaultResolve(specifier, context, defaultResolve);
    }
  } else {
    return defaultResolve(specifier, context, defaultResolve);
  }
}
```

Personally I think this approach is too clever by half, but it is terser.

addaleax commented on src/module_wrap.cc:
```diff
@@ -59,6 +60,9 @@ static const char* const EXTENSIONS[] = {
 };
 
+typedef std::unordered_set<std::string> ConditionSet;
+static std::unordered_map<std::string, ConditionSet> conditions_cache_;
```
These should probably be members of the `Enviroment` class? I don’t understand the exact semantics here, but this is definitely prone to race conditions in its current state.

(That being said, I’ll try to open a PR to move all the resolve code into JS; there is no reason why we need to complicate things by using C++, imo)

jkrems commented:
I’m not sure how others would feel if aspects of resolution are mutable at runtime and from user code. My assumption would be that NODE_ENV would be read by node on startup and then passed to loaders with runtime changes to the value being ignored.

bmeck commented on doc/api/esm.md:
```diff
@@ -1024,6 +1024,7 @@ this one, or `undefined` if this is the main entry point for the application.
  * @param {string} specifier
  * @param {object} context
  * @param {string} context.parentURL
+ * @param {string[]} context.conditions
```
It is not environmental since conditions are parameterized by callsite regarding things like `"require"`.

jkrems replied:
I don't want `resolve` hooks to need to know which conditions some node version happens to implicitly apply. I see no downside to providing a fairly complete set of conditions to the hook as long as we validate the things they pass on to the `defaultResolve` function.

guybedford replied:
Ok.

bmeck commented on src/module_wrap.cc:
```diff
@@ -1050,12 +1055,12 @@ Maybe<URL> ResolveExportsTarget(Environment* env,
       Utf8Value key_utf8(env->isolate(),
                          key->ToString(context).ToLocalChecked());
       std::string key_str(*key_utf8, key_utf8.length());
-      if (key_str == "node" || key_str == "import") {
```
I'm not sure I agree to that conclusion that we shouldn't expose a uniform system of resolution that users can tie into.

I believe the intent is to allow code like:

```mjs
export async function resolve(specifier, context, defaultResolve) {
  if (specifier.startsWith('.') || specifier.startsWith('file://') {
    // Relative or absolute specifiers get default handling
    return defaultResolve(specifier, context, defaultResolve);
  } else {
      let conditions = process.NODE_ENV = 'development' ?
        [...context.conditions, 'development'] :
        context.conditions;
      return defaultResolve(packageMetadata.name, {
        ...context,
        conditions
      });
    }
  }
}
```

If there are issues with exposing such behavior we should look into how the API has a failure to provide succinct alterations to achieve the goals users intend. I don't think the immediate response should be to hard code things and force users to re-implement logic that wouldn't share the pkg cache etc.

bmeck commented:
@jkrems I do not see this a "mutable", but as a parameterization. A single callsite will have the same resolution across the lifecycle of a node context.

jkrems commented:
> I do not see this as "mutable", but as a parameterization. A single callsite will have the same resolution across the lifecycle of a node context.

Maybe I'm not understanding the scenario but if we use the live value of `NODE_ENV` instead of passing `conditions` down, it definitely seems to be shared mutable state. For example the following (in a threaded resolver world) should be a race condition:

```js
import('some-specifier');
process.env.NODE_ENV = 'new-condition';
```

And the same call site could get different resolutions via the following:

```js
// For the case where NODE_ENV wasn't set on startup...
const doImport = async () => {
  // ...this call site gets two different resolutions for the same specifier & context
  await import('some-specifier');
};

await doImport();
process.env.NODE_ENV = 'new-condition';
await doImport();
```

Even if that doesn't match a previous definition of mutable resolution, I still think it's problematic because of the inherent risk of race conditions if mutable process-wide state can influence resolution.

bmeck commented:
@jkrems I was stating 2 fields of view that comprise my viewpoint that this is just a parameterization.

1. the loader context is potentially built when node boots up and is bootstrapped from the environment. it gets the parameters list, but that is when synchronization ends, the parameters are to build the context passed to any loader by default.
2. the loader is free to re-parameterize when delegating to a parent loader. this could change at runtime, but due to memoization it wouldn't be called multiple times per determinism guarantees of `import`.

I think the problem is thinking that `process.env.*` mutation is re-parameterizing the loader bootstrap. The loader is only being parameterized once, similar to how we snapshot CJS values once when creating ESM for them.

jkrems commented:
@bmeck I can say it more explicitly: I object to using the live `process.env.NODE_ENV` value because I don't think the loader interface for conditions should have a race condition built-in that is trivially triggered by fairly common userland code (modifying properties on `process.env`).

guybedford commented:
What if we just use `process.env.NODE_ENV` saved on loader init then?

jkrems commented:
> What if we just use process.env.NODE_ENV saved on loader init then?

That's totally possible! But that value would have to be passed to the loader hooks somehow so I don't think the fact that we extracted it from `NODE_ENV` would actually be visible to the hooks.

Another issue with _just_ passing the values from `NODE_ENV` is that it would prevent future features like using the `resolve` hook for require as well (likely as an opt-in). By including `import` in the conditions, we have a built-in way to separate the two resolution schemes. And if the hooks assume that the conditions always include `node`, it would make it impossible to reuse them for things like browser build tools. Which I think would be unfortunate.

In summary, a big +1 for parsing `NODE_ENV` on startup and using it to seed the conditions. But I think I wouldn't want the resolve hook to *just* see those values and not the other active conditions.

guybedford commented:
I've provided my full argument in https://github.com/nodejs/node/pull/31303#discussion_r389805510.

We are starting to risk some serious churn in the discussion on this PR. I'd advise everyone to focus directly on what is necessary to get this merged.

jkrems commented:
I'd like to keep discussions about how or where node may support additional conditions via flags or environment variables out of scope for this PR. The overall point to me is: I don't want `resolve` hooks to have to know about those finer details.

I'm pretty close to adding back `default` so they get the 100% complete set of conditions and make it part of the validation when calling `defaultResolve` that they don't try to call `defaultResolve` without it. I'm scared of a world where resolve hooks need to parse `process.version` and apply other magic to figure out which condition set is correct for the current resolution. I strongly belief that they should get all the information they need passed in instead of relying on implicit contracts based on multiple "well-known conditions that surely will always apply and never change".

GeoffreyBooth commented:
> so they get the 100% complete set of conditions and make it part of the validation when calling `defaultResolve`

I'm -0 on passing something like `context.supportedConditions` into `resolve`, which would tell the hook what conditions this version of Node supports; but modifying that and passing it into `defaultResolve` should have no effect (or should error). It would be the equivalent of `getFormat` receiving `context.supportedFormats` and then passing that on (or not) into `defaultGetFormat`. It's not obvious what should happen for a flow like that.

The fact that it wouldn't make sense to pass it on, though, makes me more inclined to think that it shouldn't be part of the context in the first place. If this is basically a list of constants that's true for this Node version, we could have something like `import { supportedConditions } from 'module'` and the like, and then it can be referenced within a hook but it's not part of the flow. But you don't need this list of constants to achieve the goal that I think this PR is aiming to achieve, so I don't know why we're focusing on it.

jkrems commented:
@GeoffreyBooth Can you be more explicit about your feedback? Right now this PR adds support for two feature:

1. Resolve hooks can resolve relative to the conditions that apply for this resolution request, including potential future conditions not know right now.
2. Resolve hooks can add additional conditions (e.g. "production") not recognized by node by default.
3. A resolve hook can be reused outside of node (e.g. in a bundler) where node-specific conditions may not apply or be replaced with others.

There's a very well defined meaning of adding a condition to the resolution. I'm not sure how this relates to format where that's not the case. We're adding an element to a condition set that's passed into a well-defined algorithm. This is fundamentally different from adding a new value to an enum.

I'm not aware of anything in the conditional exports spec that's ambiguous about how additional elements in the set should be interpreted. There's nothing about the algorithm that cares about which members are in the set afaik. Could you explain what you mean by "it's not obvious what should happen"? If the algorithm is ambiguous about this, we should fix that!

> If this is basically a list of constants that's true for this Node version [...]

It's not. It explicitly includes `import` which means it's relative to the exact resolution request, not the overall node version. Hooks would also be able to dynamically determine conditions that only apply to certain kinds of referring files. E.g. they could say "enable `production` only for files in `node_modules/react`" for example.

Given those two potential use cases, I'm not sure what you're trying to suggest. Are you saying that one or both shouldn't be supported? Are you suggesting a different solution?

Your suggestion seems to imply that you only want to support the first use case and you don't want to ever support `require`-resolution through `resolve` hooks. Also, the only way to use such a hook outside of node-internal resolution would be to patch `module` from my bundler which doesn't feel like a clean API. So I assume you are rejecting that use case as well (number 3 above).

It's very hard to read between the lines of your comments so it's perfectly possible that I'm misconstruing your position. Can you make a clear statement on your position on the above use cases? Without knowing what exactly you disagree with, it's hard to parse what your goal is.

GeoffreyBooth commented:
> It's very hard to read between the lines of your comments so it's perfectly possible that I'm misconstruing your position.

Well the initial comment for this PR is still blank and this PR doesn't add any documentation other than a one-line addition to the arguments for the `resolve` hook, so I've been trying to surmise from the code and this thread what this PR's goals are. Please forgive me if I made some wrong guesses.

The goals in the most recent comment seem to be more describing what this code does, but not what the use cases are. “Telling Node to load a different file for a specifier based on conditions beyond whether `require` or `import` were used” seems to be your use case, correct? And the real-world example is production/development. Right?

jkrems commented:
> “Telling Node to load a different file for a specifier based on conditions beyond whether require or import were used” seems to be your use case, correct? And the real-world example is production/development. Right?

I listed three separate use cases / constraints above. One of them is adding something like production to the condition set for the resolution of conditional exports. I also shared a complete example for this and other use cases in this thread. I don't think our documentation usually aims to contain an exhaustive list of all possible uses of a feature with complete code listings..? So I considered providing these examples in the PR discussion and to some degree tests sufficient.

jkrems commented:
* Added a note in the docs on what `conditions` is (for) and also included some basic use in the code snippet based on Geoffrey's feedback.
* Add verification to prevent custom conditions from removing `import` or `node` from the set. This should help protect some invariants as Guy pointed out ("if my exports field contains node first, that's the code that will always run in node and nothing else").

bmeck commented:
I'd like to remove the "import"/"node" restriction, if a loader is emulating a different environment it shouldn't be required to act as if it were node.

jkrems commented:
> I'd like to remove the "import"/"node" restriction, if a loader is emulating a different environment it shouldn't be required to act as if it were node.

I can see going either way on the node restriction. But for "import" it feels like it would break expectations if there's resolutions that are neither "import" nor "require".

bmeck commented:
I was particularly thinking of JSDOM style emulation which does have things like `importScripts` potentially.

jkrems commented:
> I was particularly thinking of JSDOM style emulation which does have things like importScripts potentially.

Interesting. Yes, classic script is a 3rd kind of loading system. The problem would be: What is the correct resolution behavior that node should use for them? Import-style resolution or require-style resolution? Or neither somehow? Making it an error right now to request anything but import-style resolution from the default resolve hook means we could define it later.

bmeck commented:
The exact problem is that it doesn't fall something node should handle; it would be something completely different that a loader would handle. Hence my discomfort in trying to state that it adheres to some invariant.

jkrems commented:
> The exact problem is that it doesn't fall something node should handle; it would be something completely different that a loader would handle.

The only place where we validate the conditions is before we pass them to the built-in resolution function. If a loader hook handles the conditions, they won't make it to the default hook and there wouldn't be any validation errors. Does that address your concern..?

bmeck commented:
@jkrems i could probably continually think of concrete cases where you do want the mutation of conditions, and potentially you do want to use things like the default node resolver. Combining them, I don't have a concrete use case, but I am unclear on what invariant is actually being preserved. A different approach would be to always inject those conditions in the default resolver, which seems equally bad to me as throwing. I think we could start with throwing, but the lack of ability to reuse things like the package cache seem to be concerning. I have little desire to keep the throwing behavior over time, nor to keep a potentially expanding list of conditions that node needs a matrix of passed to the resolver for it to work.

jkrems commented:
That’s totally fair. FWIW, I don’t have a strong opinion either way. I can see both the concern about surprising picks if certain built-in conditions are disabled and the added flexibility if node generally doesn’t force certain conditions beyond “default”.

I think we should continue to discuss this as the loader hooks evolve and I agree that the throwing behavior could be removed in the future.

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -1064,11 +1064,17 @@ and parent URL. The module specifier is the string in an `import` statement or
+The `conditions` property on the `context` is an array of conditions for
+[Conditional Exports][] that apply to this resolution request. They can be used
+for looking up conditional mappings elsewhere or to extend the list when calling
+the default resolution logic.
```
Thanks for adding this. @nodejs/modules-active-members does this seem understandable and good UX to you?

ljharb replied:
makes sense to me

jkrems commented:
Now with 100% less C++ thanks to @addaleax. Other notable changes included: Because line length started to become tricky with the inline type info, I switched them to JSDoc-style comments. Potentially welcome side effect is that those comments also show up in editors that support it.

jkrems commented:
Re-requested the existing approvals since the code has been ported to JS. The general idea still applies ("add a conditions argument and thread a set of strings through").

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -43,6 +47,28 @@ const {
+    return new Set(conditions);
+  }
+  return DEFAULT_CONDITIONS_SET;
+}
```
I would really prefer here that we pick one of the following two options:

a) Allow all conditions to be changed without these errors, then carefully document how conditions need to be passed through via arguments with a clear warning there about what happens if it is done wrong.
b) Do not allow these conditions to be disabled at all, and make the condition list purely additive. In that case these errors can still apply, just as all being exclusions.

I don't mind which of the above we choose - but I think it would be much clearer conceptually if we pick one path and stick with it.

guybedford replied:
That said I would still prefer treating user conditions as an extra layer on top of the Node.js internal conditions that apply (no repetition necessary, so no errors above), but am happy to just see the errors disabled too with a careful docs comment.

bmeck replied:
I think both a) and b) have the potential for future breakage if we add conditions to the default resolver, the same for any custom conditions we may pass in. I think no matter the way forward we need to admit that breakage and thus errors can occur. I like the current approach since it carves out a small subset of known desired conditions that we have talked about and are in resolution semantics already.

guybedford replied:
@bmeck I don't think this argument holds for loaders since it is always superseded in effect by the base breaking scenario for any default resolver.

**Adding any new condition** to the default resolver is a breaking change, regardless of this loader hook. What really counts is how established user patterns are - kind of like how browsers could add `Array.prototype.includes`  as non-breaking but not `Array.prototype.contains`.

With the current approach, we risk new conditions actually throwing loaders that hard code conditions. Eg a loader that is experimenting with a "production" condition breaks when that condition is made to throw since it is a core condition. This seems detrimental to forward compatibility of new experimentation.

With both (a) and (b) a loader that experiments with new conditions then layering into that condition being brought into core continues to function fine.

jkrems replied:
I think erring on the side of being permissive (and not forcing conditions) would be the correct direction. If we can agree on a direction here, happy to seal the deal and make the change.

My preferred solution would be "node accepts the conditions passed and will only implicitly add default". Testing tools like JSDOM have good reasons *not* to want the node-specific implicit conditions. I was fine adding the current exceptions on the premise that removing them in follow-up PRs is easier than adding them.

guybedford replied:
Ok @jkrems if that is your preference I can get behind it with these errors removed but a very careful documentation comment about this, and specifically the need for loader to pass these conditions and risks if they don't.

jkrems replied:
Removed the exception and added a sentence to the docs.

guybedford replied:
Looks great, thanks.

bmeck commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -43,6 +47,28 @@ const {
 } = require('internal/errors').codes;
 
+const DEFAULT_CONDITIONS = ObjectFreeze(['node', 'import']);
+const DEFAULT_CONDITIONS_SET = new Set(['node', 'import']);
```
```suggestion
const DEFAULT_CONDITIONS_SET = new Set(DEFAULT_CONDITIONS);
```

ljharb replied:
```suggestion
const DEFAULT_CONDITIONS_SET = new SafeSet(DEFAULT_CONDITIONS);
```

bmeck commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -357,11 +388,12 @@ function resolveExportsTarget(
     for (let i = 0; i < keys.length; i++) {
       const key = keys[i];
-      if (key === 'node' || key === 'import' || key === 'default') {
+      if (key === 'default' || conditions.has(key)) {
```
```suggestion
      if (key === 'default' || primordials.SetPrototypeHas(conditions, key)) {
```

ljharb replied:
this isn't needed if `conditions` is a SafeSet, i think - unless `conditions` is coming from user loaders, in which case, this is definitely needed

ljharb commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -2,11 +2,14 @@
   ObjectGetOwnPropertyNames,
   ObjectPrototypeHasOwnProperty,
   SafeMap,
+  Set,
```
```suggestion
  SafeSet,
```

bmeck replied:
I don't think we want to directly expose `SafeSet` to users.

jkrems replied:
To the user, we only expose the array. The `Set` is only passed into internal functions.

bmeck replied:
k

ljharb commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -43,6 +47,28 @@ const {
+      throw new ERR_INVALID_ARG_VALUE('conditions', conditions,
+                                      'has to contain "node"');
+    }
+    return new Set(conditions);
```
```suggestion
    return new SafeSet(conditions);
```

jkrems commented:
Switched to `SafeSet`.

guybedford commented on doc/api/esm.md:
```diff
@@ -1053,11 +1053,19 @@ and parent URL. The module specifier is the string in an `import` statement or
+for looking up conditional mappings elsewhere or to modify the list when calling
+the default resolution logic. Since `resolve` hooks have the ability to remove
+matching of conditions that ship with node like `node`, this can lead to files
+being executed that were meant for other JavaScript environments.
```
Can we split this up into two paragraphs, with the second explicitly calling out the actionable part:

```suggestion
The `conditions` property on the `context` is an array of conditions for
[Conditional Exports][] that apply to this resolution request. They can be used
for looking up conditional mappings elsewhere or to modify the list when calling
the default resolution logic.

_The Node.js default conditions `"node"` and `"import"` will always be in the default
conditions list and **must** be passed through in the returned conditions list to
ensure Node.js-compatible resolution logic._
```

jkrems replied:
Agreed on making it two paragraphs. But I think that 2nd paragraph is too strong. E.g. I would want to reserve the right to pass `require` in the future instead of `import`. And there should be be an if before the **must**, otherwise it sounds like an unqualified **must** (which it isn't). How about:

> The Node.js [default conditions listed above][link-here] will always be in the `context.conditions` list passed to the hook. _If the hook wants to ensure Node.js-compatible resolution logic, the conditions list **must** be passed through to the `defaultResolve` function._

guybedford replied:
Looks great - just be specific about the condition list - `"all items from this default condition list must be passed through..."`.

jkrems replied:
Updated with the new text. Less cursive because multiple lines in markdown with cursive are annoying to manage. Felt okay because it makes the **must** stand out more imo.

DerekNonGeneric commented on doc/api/esm.md:
```diff
@@ -1053,11 +1053,22 @@ and parent URL. The module specifier is the string in an `import` statement or
  * @param {string} specifier
  * @param {object} context
  * @param {string} context.parentURL
+ * @param {string[]} context.conditions
```
I wish this line read the following.

```javascript
@param {Array<string>} context.conditions
```

That would make this annotation compatible w/ Closure.

jkrems replied:
I think for practical purposes I'd rather use syntax familiar to Typescript devs since *generally* I would expect the number of people using VSCode and/or TypeScript to be a lot bigger than the number of people actively using Closure Compiler.

DerekNonGeneric commented on doc/api/esm.md:
````diff
@@ -1053,11 +1053,22 @@ and parent URL. The module specifier is the string in an `import` statement or
 ```js
 /**
  * @param {string} specifier
  * @param {object} context
````
Non-null object in Closure:
```javascript
@param {!(Object)}
```

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -43,6 +45,16 @@ const {
+const DEFAULT_CONDITIONS_SET = new SafeSet(DEFAULT_CONDITIONS);
+
+function getConditionsSet(conditions) {
+  if (conditions !== DEFAULT_CONDITIONS && ArrayIsArray(conditions)) {
```
We should throw an error ERR_INVALID_ARG_TYPE for the loader if it is not an array or undefined / null.

bmeck replied:
Should we validate that all items are strings? I think default coercion to string is probably fine for all existing types / cannot imagine non-string items being useful right now.

jkrems replied:
I don't think coercion is happening right now (if I understand your assumption correctly). Any non-string element in the array should be ignored right now since we're using set membership tests.

jkrems replied:
I added a check that throws when it's a non-array.

Himself65 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -18,7 +18,10 @@ const { validateString } = require('internal/validators');
-const { defaultResolve } = require('internal/modules/esm/resolve');
+const {
+  defaultResolve,
+  DEFAULT_CONDITIONS,
```
I would like to remove `DEFAULT_CONDITIONS` on `module.exports` and use

```js
function moduleResolve(specifier, base, conditions = ['import','node']) {
  // ...
}
```

jkrems replied:
The issue is that we need to pass those into a userland function, not just into our own default implementation. So I don't think that can be replaced with default arguments. Unless of course I'm missing something about the suggestion. :)

jkrems commented:
There's one C++ test that seems to fail on linuxone. I look into it briefly but I don't see a clear reason why it would fail (or what the failure really means). I'm surprised that this PR would affect low-level worker interactions but can't rule it out.

addaleax commented:
@jkrems See #32563 – I just landed a fix for that. :) Is this still ready to land, apart from waiting for a green CI?

jkrems commented:
> Is this still ready to land, apart from waiting for a green CI?

Yes! This should be good to go, assuming CI comes back green.

addaleax commented:
Landed in 9129ab1996251245b06d8df599634d31c75849ca
